src/lsm/forest_fuzz.zig

```diff
@@ -262,8 +277,77 @@ const Environment = struct {
 
     fn apply(env: *Environment, fuzz_ops: []const FuzzOp) !void {
         // The forest should behave like a simple key-value data-structure.
-        // We'll compare it to a hash map.
-        var model = std.hash_map.AutoHashMap(u128, Account).init(allocator);
+        const Model = struct {
+            checkpointed: KVType, // represents persistent state
+            log: LogType, // represents in-memory state
+
+            const KVType = std.hash_map.AutoHashMap(u128, Account);
+            const LogEntry = struct { op: u64, account: Account };
+            const LogType = std.fifo.LinearFifo(LogEntry, .Dynamic);
+            const Model = @This();
+
+            pub fn init() Model {
+                return .{
+                    .checkpointed = KVType.init(allocator),
+                    .log = LogType.init(allocator),
+                };
+            }
+
+            pub fn deinit(model: *Model) void {
+                model.checkpointed.deinit();
+                model.log.deinit();
+            }
+
+            pub fn put_account(model: *Model, account: *const Account, op: u64) !void {
+                try model.log.writeItem(.{ .op = op, .account = account.* });
+            }
+
+            pub fn get_account(model: *const Model, id: u128) ?Account {
+                var latest_op: ?u64 = null;
+                const log_size = model.log.readableLength();
+                var log_left = log_size;
+                while (log_left > 0) : (log_left -= 1) {
+                    const entry = model.log.peekItem(log_left - 1); // most recent first
+                    if (latest_op == null) {
+                        latest_op = entry.op;
+                    }
+
+                    assert(latest_op.? >= entry.op);
+
+                    if (entry.account.id == id) {
+                        return entry.account;
+                    }
+                }
+                return model.checkpointed.get(id);
+            }
+
+            pub fn checkpoint(model: *Model, op: u64) !void {
+                const checkpointable = blk: {
+                    if (op > constants.lsm_batch_multiple) {
+                        break :blk op - (op % constants.lsm_batch_multiple) - 1;
+                    } else {
+                        break :blk 0;
+                    }
+                };
+                const log_size = model.log.readableLength();
+                var log_left = log_size;
+                while (log_left > 0) : (log_left -= 1) {
+                    const entry = model.log.peekItem(log_size - log_left);
```

Comment:
Since you use `log_size - log_left` everywhere this seems to be equivalent to:

``` zig
var log_index: usize = 0;
while (log_index < log_size) : (log_index += 1) {
    const entry = model.log.peekItem(log_index);
    ...
```

Reply:
Good catch! Will change